File: 0050-pow-x-n/50.py
class Solution:
    def myPow(self, x: float, n: int) -> float:
        # Binary Exponentiation -- T.C: O(log N),  S.C: O(1)
        if n < 0:
            x = 1 / x
            n = -n

        res = 1.0
        while n:
            if n % 2:
                res *= x
            x *= x
            n //= 2

        return res

File: 0050-pow-x-n/50.rs
impl Solution {
    pub fn my_pow(x: f64, n: i32) -> f64 {
        // Binary Exponentiation -- T.C: O(log N), S.C: O(1)
        let mut x = x;
        let mut N: i64 = n as i64;
        
        if n < 0 {
            N = -N;
            x = 1.0 / x;
        }

        let mut res: f64 = 1.0;
        while N > 0 {
            if N & 1 != 0 {
                res *= x;
            }
            x *= x;
            N = N / 2;
        }
        res
    }
}
